### PR TITLE
Sql Task. AlexErm

### PR DESCRIPTION
**Pull request by alexErm0**
https://travis-ci.com/github/alexErm0/db-assignments/builds/178248318

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -91,7 +133,17 @@ async function task_1_5(db) {
...
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           Products.ProductName,
+           Categories.CategoryName,
+           Suppliers.CompanyName as SupplierCompanyName
+        FROM ((Products
+            NATURAL JOIN Categories)
```
Comment (AlexMokin) [inline task/sql-tasks.js]: avoid using natural join

File: task/sql-tasks.js [hunk 2]
```diff
@@ -106,7 +158,20 @@ async function task_1_6(db) {
  *
  */
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           first.EmployeeID as "EmployeeId",
+           CONCAT(first.FirstName, ' ', first.LastName) AS "FullName",
+           CASE
```
Comment (AlexMokin) [inline task/sql-tasks.js]: check ifnull and coalesce funcitons

File: task/sql-tasks.js [hunk 3]
```diff
@@ -118,7 +183,21 @@ async function task_1_7(db) {
  *
  */
 async function task_1_8(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           Categories.CategoryName,
+           COUNT(Products.CategoryID) as TotalNumberOfProducts
+        FROM Categories
+        NATURAL JOIN Products
```
Comment (AlexMokin) [inline task/sql-tasks.js]: avoid using natural join

File: task/sql-tasks.js [hunk 4]
```diff
@@ -118,7 +183,21 @@ async function task_1_7(db) {
...
  */
 async function task_1_8(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           Categories.CategoryName,
+           COUNT(Products.CategoryID) as TotalNumberOfProducts
+        FROM Categories
+        NATURAL JOIN Products
+        WHERE Products.CategoryID IN (
```
Comment (AlexMokin) [inline task/sql-tasks.js]: what are you trying to filter out here?

File: task/sql-tasks.js [hunk 5]
```diff
@@ -178,7 +294,16 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           COUNT(Continued) AS TotalOfCurrentProducts,
+           SUM(Continued) AS TotalOfDiscontinuedProducts
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count records, it better shows intent

 it's better to write this query like this
notice that from-clause is optional.



```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 6]
```diff
@@ -200,7 +333,23 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM(IF(OrderDate LIKE "1997-01-%", 1, 0)) as January,
+            SUM(IF(OrderDate LIKE "1997-02-%", 1, 0)) as February,
+            SUM(IF(OrderDate LIKE "1997-03-%", 1, 0)) as March,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count records, it better shows intent

File: task/sql-tasks.js [hunk 7]
```diff
@@ -200,7 +333,23 @@ async function task_1_14(db) {
...
+            SUM(IF(OrderDate LIKE "1997-04-%", 1, 0)) as April,
+            SUM(IF(OrderDate LIKE "1997-05-%", 1, 0)) as May,
+            SUM(IF(OrderDate LIKE "1997-06-%", 1, 0)) as June,
+            SUM(IF(OrderDate LIKE "1997-07-%", 1, 0)) as July,
+            SUM(IF(OrderDate LIKE "1997-08-%", 1, 0)) as August,
+            SUM(IF(OrderDate LIKE "1997-09-%", 1, 0)) as September,
+            SUM(IF(OrderDate LIKE "1997-10-%", 1, 0)) as October,
+            SUM(IF(OrderDate LIKE "1997-11-%", 1, 0)) as November,
+            SUM(IF(OrderDate LIKE "1997-12-%", 1, 0)) as December
```
Comment (AlexMokin) [inline task/sql-tasks.js]: check year in where clause

File: task/sql-tasks.js [hunk 8]
```diff
@@ -224,7 +381,16 @@ async function task_1_16(db) {
  *
  */
 async function task_1_17(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           CategoryName,
+           AVG(Products.UnitPrice) AS AvgPrice
+        FROM Categories
+        NATURAL JOIN Products
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not use natural join

File: task/sql-tasks.js [hunk 9]
```diff
@@ -248,7 +422,20 @@ async function task_1_18(db) {
...
+            Customers.CustomerID,
+            Customers.CompanyName,
+            SUM(OrderDetails.UnitPrice * OrderDetails.Quantity)
+                as "TotalOrdersAmount, $"
+        FROM ((Customers
+            NATURAL JOIN Orders)
+            NATURAL JOIN OrderDetails)
+        GROUP BY CustomerID
+        HAVING SUM(OrderDetails.UnitPrice * OrderDetails.Quantity) > 10000
+        ORDER BY SUM(OrderDetails.UnitPrice * OrderDetails.Quantity) DESC, CustomerID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not duplicate formula in order by just use column name

File: task/sql-tasks.js [hunk 10]
```diff
@@ -270,7 +469,17 @@ async function task_1_20(db) {
  * @return {array}
  */
 async function task_1_21(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            OrderID,
+            SUM(UnitPrice * Quantity) as "Maximum Purchase Amount, $"
+        FROM Orders
+        NATURAL JOIN OrderDetails
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not use natural join

File: task/sql-tasks.js [hunk 11]
```diff
@@ -281,7 +490,25 @@ async function task_1_21(db) {
...
+        FROM Customers
+            INNER JOIN Orders ON Customers.CustomerID = Orders.CustomerID
+            INNER JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+            INNER JOIN Products ON Products.ProductID = OrderDetails.ProductID
+        WHERE OrderDetails.UnitPrice IN (
+            SELECT MAX(UnitPrice)
+            FROM Customers cust
+                INNER JOIN Orders ON Customers.CustomerID = Orders.CustomerID
+                INNER JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+            WHERE cust.CompanyName = Customers.CompanyName
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.

**Review by AlexMokin — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by alexErm0** [comment]
https://travis-ci.com/github/alexErm0/db-assignments/builds/182288443

File: task/sql-tasks.js [hunk 12]
```diff
@@ -91,7 +133,17 @@ async function task_1_5(db) {
  *
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           Products.ProductName,
+           Categories.CategoryName,
+           Suppliers.CompanyName as SupplierCompanyName
+        FROM ((Products
```
Comment (AlexMokin) [inline task/sql-tasks.js]: Remove redundant parentheses

File: task/sql-tasks.js [hunk 13]
```diff
@@ -200,7 +319,23 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            COUNT(IF(OrderDate LIKE "1997-01-%", 1, NULL)) as January,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: try to implement it like this

```
Select
Count(condition) as Jan
....
From Orders
Where year = 1997
```

File: task/sql-tasks.js [hunk 14]
```diff
@@ -260,7 +433,19 @@ async function task_1_19(db) {
...
+    let result = await db.query(`
+        SELECT 
+            Employees.EmployeeID,
+            CONCAT(FirstName, ' ', LastName) AS "Employee Full Name",
+            SUM(UnitPrice * Quantity) as "Amount, $"
+        FROM Employees
+            INNER JOIN Orders ON Employees.EmployeeID = Orders.EmployeeID
+            INNER JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+        GROUP BY Employees.EmployeeID
+        ORDER BY SUM(UnitPrice * Quantity) DESC
```
Comment (AlexMokin) [inline task/sql-tasks.js]: du not duplicate formula

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by AlexMokin** [comment]
accepted

**Pull request closed without merging**